src/image_surface_png.rs

```diff
@@ -0,0 +1,110 @@
+// Copyright 2016, The Gtk-rs Project Developers.
+// See the COPYRIGHT file at the top-level directory of this distribution.
+// Licensed under the MIT license, see the LICENSE file or <http://opensource.org/licenses/MIT>
+
+use std::slice;
+use std::mem::transmute;
+use std::io;
+use std::io::{Read, Write, Error, ErrorKind};
+use std::process;
+use std::thread;
+
+use libc::{c_void, c_uint};
+
+use glib::translate::*;
+use ffi;
+use ffi::enums::Status;
+use ImageSurface;
+
+
+struct CallbackGuard;
+
+impl Drop for CallbackGuard {
+    fn drop(&mut self) {
+        if thread::panicking() {
+            process::exit(101);
+        }
+    }
+}
+
+struct ReadEnv<R: Read> {
+    reader: R,
+    error: Option<Error>,
+}
+
+unsafe extern "C" fn read_func<R: Read>(closure: *mut c_void, data: *mut u8, len: c_uint) -> Status {
+    let _cbguard = CallbackGuard;
+    let read_env: &mut ReadEnv<R> = &mut *(closure as *mut ReadEnv<R>);
+    let mut buffer = slice::from_raw_parts_mut(data, len as usize);
+
+    // Lifted from std::io::Read::read_exact
+    while !buffer.is_empty() {
+        match read_env.reader.read(buffer) {
+            Ok(0) => break,
+            Ok(n) => { let tmp = buffer; buffer = &mut tmp[n..]; }
+            Err(ref e) if e.kind() == ErrorKind::Interrupted => {}
+            Err(e) => {
+                read_env.error = Some(e);
+                return Status::ReadError;
+            }
+        }
+    }
+    if !buffer.is_empty() {
+        read_env.error = Some(Error::new(ErrorKind::Other, "failed to fill whole buffer"));
+        Status::ReadError
+    } else {
+        Status::Success
+    }
+}
+
+struct WriteEnv<W: Write> {
+    writer: W,
+    error: Option<Error>,
+}
+
+unsafe extern "C" fn write_func<W: Write>(closure: *mut c_void, data: *mut u8, len: c_uint) -> Status {
+    let _cbguard = CallbackGuard;
+    let write_env: &mut WriteEnv<W> = &mut *(closure as *mut WriteEnv<W>);
+    let buffer = slice::from_raw_parts(data, len as usize);
+    match write_env.writer.write_all(buffer) {
+        Ok(()) => Status::Success,
+        Err(error) => {
+            write_env.error = Some(error);
+            Status::WriteError
+        },
+    }
+}
+
+
+#[derive(Debug)]
+pub enum PngError { Cairo(Status), Io(Error) }
```

Comment:
There's nothing tying this enum to png. I'd rather it had a more generic name e.g. `IoError` and live in its own module, say `error`.